### PR TITLE
MFA 

### PR DESCRIPTION
**Pull request by MrFlashAccount**
### Pull Request Description

This PR is based on #10844 and #10775, so merge/review them in order.

TLDR: This PR adds 2FA support for Login flow

Closes: enso-org/cloud-v2#1428

Demo: 

https://github.com/user-attachments/assets/9eaf66c9-ac28-4b5b-a066-a4d8573d4eaf

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

**Comment by PabloBuchu** [comment]
That looks really great! 

1. When I am being asked for the code and I provide the wrong one I am being redirected to login screen. Can we stay with the provide code screen but display `Code mismatch` error?
2. How we can provide a recovery path for this (preferably in separate branch / issue)

**Comment by MrFlashAccount** [comment]
> When I am being asked for the code and I provide the wrong one I am being redirected to login screen. Can we stay with the provide code screen but display Code mismatch error?

We redirect back to login screen only if the cognito responds with `LoginTimeout` or when a user made too much attempts to enter a code.

> How we can provide a recovery path for this (preferably in separate branch / issue)

You mean to provide a way to disable TOTP or use recovery codes? AFAIC cognito doesn't provide that out of the box, but we can provide alternative ways to get access to the account - using SMS 2fa or by requesting support.

**Comment by MrFlashAccount** [comment]
> When I am being asked for the code and I provide the wrong one I am being redirected to login screen. Can we stay with the provide code screen but display Code mismatch error?

Can't repro, can you provide more details on how to reproduce that?

**Comment by PabloBuchu** [comment]
@MrFlashAccount all is good. lets try to clean and merge yours PRs ✅

**Comment by PabloBuchu** [comment]
Unfortunatelly it stopped working. When I want to sign in I am getting

Screenshot 2024-09-10 at 14 49 55

File: app/dashboard/src/layouts/Settings/settingsData.tsx [hunk 1]
```diff
@@ -124,6 +125,23 @@ export const SETTINGS_TAB_DATA: Readonly<Record<SettingsTabType, SettingsTabData
           },
         ],
       },
+      {
+        nameId: 'setupTwoFaSettingsSection',
```
Comment (somebody1234) [inline app/dashboard/src/layouts/Settings/settingsData.tsx]: nit: maybe 2Fa or 2fa instead of TwoFa?

File: app/dashboard/src/modules/payments/components/PlanSelector/components/PlanSelectorDialog.tsx [hunk 2]
```diff
@@ -145,7 +155,7 @@ export function PlanSelectorDialog(props: PlanSelectorDialogProps) {
                   form={form}
                   name="period"
                   // eslint-disable-next-line @typescript-eslint/no-magic-numbers
-                  items={[12, 36]}
+                  items={[1, 12, 36]}
```
Comment (somebody1234) [inline app/dashboard/src/modules/payments/components/PlanSelector/components/PlanSelectorDialog.tsx]: monthly is back?

File: app/dashboard/src/providers/AuthProvider.tsx [hunk 3]
```diff
@@ -92,7 +92,13 @@ interface AuthContextType {
   readonly setUsername: (username: string) => Promise<boolean>
   readonly signInWithGoogle: () => Promise<boolean>
   readonly signInWithGitHub: () => Promise<boolean>
-  readonly signInWithPassword: (email: string, password: string) => Promise<void>
+  readonly signInWithPassword: (
+    email: string,
+    password: string,
+  ) => Promise<{
```
Comment (somebody1234) [inline app/dashboard/src/providers/AuthProvider.tsx:98]: ideally this should be its own interface IMO so that it shows up as its name in hover tooltips

File: app/dashboard/src/modals/SetupOrganizationAfterSubscribe.tsx [hunk 4]
```diff
@@ -163,6 +163,7 @@ export const SET_ORGANIZATION_NAME_FORM_SCHEMA = (getText: GetText) =>
     name: ariaComponents.Form.schema
       .string()
       .min(1, getText('arbitraryFieldRequired'))
+      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
```
Comment (somebody1234) [inline app/dashboard/src/modals/SetupOrganizationAfterSubscribe.tsx]: is this necessary?

**Review by somebody1234 — APPROVED** [review]
CR :white_check_mark:

**Comment by somebody1234** [comment]
so i guess we still want to support oss builds, so i guess we will still need to support cloud endpoints being missing (null) both in the codebase and the environment variable type definitions

**Comment by MrFlashAccount** [comment]
> so i guess we still want to support oss builds, so i guess we will still need to support cloud endpoints being missing (null) both in the codebase and the environment variable type definitions

We cut this functionally a while ago when we made signing up a required step. I agree it's a clash and we must discuss it with James. 
But, in general, giving prod env variables to the users so they can build the app themselves might be the solution.

File: app/dashboard/src/modules/payments/components/PlanSelector/components/PlanSelectorDialog.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/dashboard/src/modules/payments/components/PlanSelector/components/PlanSelectorDialog.tsx]: No! I guess it went back after a rebase. Thanks for noticing!

File: app/dashboard/src/providers/AuthProvider.tsx (hunk 3 above)
Comment (MrFlashAccount) [reply app/dashboard/src/providers/AuthProvider.tsx:98]: I think it's better to see the keys in tooltips, isn't it? Also, extracting to a separate interface might be a good idea for reusability, but I'm, at the moment, not sure we need to do that

File: app/dashboard/src/modals/SetupOrganizationAfterSubscribe.tsx (hunk 4 above)
Comment (MrFlashAccount) [reply app/dashboard/src/modals/SetupOrganizationAfterSubscribe.tsx]: Not at all
